ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs

```diff
@@ -633,25 +659,33 @@ modifyState VolatileDBEnv{_dbHasFS = hasFS :: HasFS m h, ..} action = do
       -- In case of an exception, close the DB for safety.
       ExitCaseException _ex -> do
         putMVar _dbInternalState VolatileDbClosed
-        closeOpenHandle mst
+        closeAllHandles hasFS mst
       -- In case of success, update to the newest state.
       ExitCaseSuccess (Right (newState, _)) ->
         putMVar _dbInternalState (VolatileDbOpen newState)
       -- In case of an error (not an exception), close the DB for safety.
       ExitCaseSuccess (Left _) -> do
         putMVar _dbInternalState VolatileDbClosed
-        closeOpenHandle mst
+        closeAllHandles hasFS mst
 
     mutation :: OpenOrClosed blockId h
              -> m (InternalState blockId h, r)
     mutation VolatileDbClosed          = throwError $ UserError ClosedDBError
     mutation (VolatileDbOpen oldState) = action hasFS oldState
 
-    -- TODO what if this fails?
-    closeOpenHandle :: OpenOrClosed blockId h -> m ()
-    closeOpenHandle VolatileDbClosed                    = return ()
-    closeOpenHandle (VolatileDbOpen InternalState {..}) =
-      wrapFsError hasFsErr _dbErr $ hClose _currentWriteHandle
+closeAllHandles :: forall blockId m h. (HasCallStack, IOLike m)
+                => HasFS m h
+                -> OpenOrClosed blockId h
+                -> m ()
+closeAllHandles hasFS = closeWithState
+  where
+    HasFS{..} = hasFS
+
+    closeWithState VolatileDbClosed = return ()
+    closeWithState (VolatileDbOpen InternalState {..}) = do
+      hClose _currentWriteHandle
+      mapM_ (hClose . FileInfo.getHandle) $ Index.elems _currentMap
```

Comment:
> What if closing the first handle fails?
> 
> I think it would be good to use a `ResourceRegistry`for the handles. When closing the `VolatileDB`, you simply close the registry, which is already [able](https://github.com/input-output-hk/ouroboros-network/blob/e9430d56c1378ff31d05ddd63b7f964a6ff2c293/ouroboros-consensus/src/Ouroboros/Consensus/Util/ResourceRegistry.hs#L539) to deal with exceptions thrown in the clean-up actions.

I have some questions about using `ResourceRegistry` for this:
- Will it be the same registry as the `cdbRegistry` of `ChainDbArgs` or a new one should be created for the VolatileDB? I guess the latter or else closing the registry from the VolatileDB would also release resources of the ChainDB.
- Even if VolatileDB has its own dedicated `ResourceRegistry`, it's possible to close and reopen the db, but the registry can't reopen. Also `closeRegistry` is only exported for tests and has to be used by the same thread that opened it. However in the VolatileDB I can't make sure it's the same thread. Maybe a new function `relaseAll` or `cleanRegistry` should be added for this.
- `ResourceKey` is not exported from `Ouroboros.Consensus.Util.ResourceRegistry`. Is there any reason for this or should I just export it? (it's used to call release).

Reply:
> > What if closing the first handle fails?
> > I think it would be good to use a `ResourceRegistry`for the handles. When closing the `VolatileDB`, you simply close the registry, which is already [able](https://github.com/input-output-hk/ouroboros-network/blob/e9430d56c1378ff31d05ddd63b7f964a6ff2c293/ouroboros-consensus/src/Ouroboros/Consensus/Util/ResourceRegistry.hs#L539) to deal with exceptions thrown in the clean-up actions.
> 
> I have some questions about using `ResourceRegistry` for this:
> 
>     * Will it be the same registry as the `cdbRegistry` of `ChainDbArgs` or a new one should be created for the VolatileDB? I guess the latter or else closing the registry from the VolatileDB would also release resources of the ChainDB.

Indeed a new one for the VolatileDB.

> 
>     * Even if VolatileDB has its own dedicated `ResourceRegistry`, it's possible to close and reopen the db, but the registry can't reopen. Also `closeRegistry` is only exported for tests and has to be used by the same thread that opened it. However in the VolatileDB I can't make sure it's the same thread. Maybe a new function `relaseAll` or `cleanRegistry` should be added for this.

Hmm, can't you create a new registry when reopening? Good point about it needing to be the same thread :thinking: 
> 
>     * `ResourceKey` is not exported from `Ouroboros.Consensus.Util.ResourceRegistry`. Is there any reason for this or should I just export it (opaque)?

Yeah, I think it was just forgotten.

Reply:
> Hmm, can't you create a new registry when reopening?

Maybe it is possible, but in this case I have to use the `openRegistry`/`closeRegistry` combinators instead of `withRegistry`. I think the VolatileDB should not create and close the registry: it should be given to it as an argument when opening, like in mempool. The registry should be created when the node starts like in `Ouroboros.Consensus.Node`. That's why I was thinking about a `releaseAll` function. 

>  Good point about it needing to be the same thread :thinking:

A `releaseAllunsafe` function which does not check about threads would also fix this. There are already other locks in the VolatileDB which protect against, for example, trying to use a handles which is already closed.